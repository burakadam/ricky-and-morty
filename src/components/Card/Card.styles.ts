import styled from '@emotion/styled';
import { Link } from 'react-router-dom';

const Nav = styled(Link)`
  position: absolute;
  height: 100%;
  width: 100%;
  left: 0;
  right: 0;
  z-index: 10;
`;

export { Nav };
