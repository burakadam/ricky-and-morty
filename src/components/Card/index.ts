import type { Character } from '@/gql/graphql';

export interface ICardItemProps {
  character: Character;
}

export * from './Card';
