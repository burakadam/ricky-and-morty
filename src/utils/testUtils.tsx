import type { PreloadedState } from '@reduxjs/toolkit';
import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import type { PropsWithChildren } from 'react';
import React from 'react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '@/store/store';
import { setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/no-shadow
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function Wrapper({ children }: { children: React.ReactElement }) {
  return <>{children}</>;
}

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
