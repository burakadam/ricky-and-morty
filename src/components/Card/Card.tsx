import { ImageListItem, ImageListItemBar } from '@mui/material';

import ROUTES from '@/constants/routes';

import type { ICardItemProps } from '.';
import * as S from './Card.styles';

const Card = ({ character }: ICardItemProps) => {
  const { image, name, id, ...rest } = character;
  const subtitle = Object.values(rest).join(' / ');

  return (
    <ImageListItem>
      <S.Nav to={`${ROUTES.CHARACTERS}/${id}`} />
      <img src={image ?? ''} alt={name ?? 'Character'} loading="lazy" />
      <ImageListItemBar title={name} subtitle={<p>{subtitle}</p>} />
    </ImageListItem>
  );
};

export { Card };
