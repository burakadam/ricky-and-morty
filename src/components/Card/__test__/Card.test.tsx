/* eslint-disable testing-library/no-node-access */
import { screen } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';

import type { Character } from '@/gql/graphql';
import { renderWithProviders } from '@/utils/testUtils';

import { Card } from '../Card';

const SAMPLE: Character = {
  __typename: 'Character',
  name: 'Alien Googah',
  id: '13',
  image: 'https://rickandmortyapi.com/api/character/avatar/13.jpeg',
  status: 'unknown',
  gender: 'unknown',
  type: 'type',
  species: 'Alien',
  episode: []
};

const FEATURES = `${SAMPLE.status} / ${SAMPLE.gender} / ${SAMPLE.type} / ${SAMPLE.species} /`;

describe('Character list card', () => {
  let container: HTMLElement;

  beforeEach(() => {
    // eslint-disable-next-line testing-library/no-render-in-setup
    ({ container } = renderWithProviders(
      <Router>
        <Card character={SAMPLE} />
      </Router>
    ));
  });

  it('render component correctly', () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it('contain list of status, gender, type and species', () => {
    expect(screen.getByText(FEATURES)).toBeInTheDocument();
  });

  it('contain name of the image url', () => {
    const image = container.querySelector('img') as HTMLImageElement;
    expect(image.src).toEqual(SAMPLE.image);
  });

  it('contain url with character id', () => {
    const link = container.querySelector('a') as HTMLAnchorElement;
    expect(link.href).toEqual(`http://localhost/characters/${SAMPLE.id}`);
  });
});
