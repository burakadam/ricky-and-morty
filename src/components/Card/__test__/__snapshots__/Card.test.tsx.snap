// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character list card render component correctly 1`] = `
<li
  class="MuiImageListItem-root css-l086p4-MuiImageListItem-root"
  style="height: auto; grid-column-end: span 1; grid-row-end: span 1;"
>
  <a
    class="css-1lu6ooa"
    href="/characters/13"
  />
  <img
    alt="Alien Googah"
    class="MuiImageListItem-img"
    loading="lazy"
    src="https://rickandmortyapi.com/api/character/avatar/13.jpeg"
  />
  <div
    class="MuiImageListItemBar-root MuiImageListItemBar-positionBottom css-ztrqiq-MuiImageListItemBar-root"
  >
    <div
      class="MuiImageListItemBar-titleWrap MuiImageListItemBar-titleWrapBottom css-1lizpae-MuiImageListItemBar-titleWrap"
    >
      <div
        class="MuiImageListItemBar-title css-dasnyc-MuiImageListItemBar-title"
      >
        Alien Googah
      </div>
      <div
        class="MuiImageListItemBar-subtitle css-hk6gwf-MuiImageListItemBar-subtitle"
      >
        <p>
          unknown / unknown / type / Alien / 
        </p>
      </div>
    </div>
  </div>
</li>
`;
